src/ast/semantic_analyser.cpp

```diff
@@ -2318,6 +2324,41 @@ int SemanticAnalyser::create_maps(bool debug)
     failed_maps += is_invalid_map(bpftrace_.perf_event_map_->mapfd_);
   }
 
+  size_t max_zero_buffer_size_ = 0;
+
+  if (needs_fmtstr_map_)
+  {
+    std::string map_ident = "fmtstr";
+
+    size_t printf_struct_size = sizeof(size_t) + max_fmtstr_args_size_;
+    SizedType type = SizedType(Type::fmtstr, printf_struct_size);
```

Comment:
Do we really need a a Type::fmtstr? Isn't a format string still a string?

Reply:
this was inherited from @mmarchini's https://github.com/iovisor/bpftrace/pull/750.

fmtstr is the structure composed by `CreateFormatString()` (e.g. printf, join) and emitted as a perf event.

I think now that str is handled the same way by map.cpp, fmtstr could be changed to str without breaking anything. but that risks being disingenuous; it's not a "char array of length n" — it's a struct (whose length depends on the lengths of the operands to printf()).

Since it's a struct not a string: I think it still makes sense to continue calling it fmtstr.

Reply:
removed to make the review smaller.

Reply:
Hmm, I'm not fully convinced on this point. I'm not really seeing in which cases the type system would need to disambiguate between a format string and a regular string. For `printf`, `system`, etc, the semantic analyser + codegen knows it's a format string based on the position of the argument. I'm not really sure why we would need additional help to know it's a format string.

For example in C, a format string is still a `char *`. `sprintf` takes a `char *` + some additional arguments and "creates" and new `char *`.

My concern with new types is that each new type creates edge cases and strange interactions between types. For example, we have a `count` type that cannot be compared to `int` type. (That one has a technical reason but it'd be awesome if `count` would just be an `int`).

Is there something I'm missing?

Reply:
I suspect the entire motivation of the original code was "the type system determines what type of Map you will get from map.cpp".

the additional changes in my PR (since the original printf RFC https://github.com/iovisor/bpftrace/pull/750) ensure that `Types::str` gets a per-CPU array from map.cpp, so on that basis I think we no longer need fmtstr — regular str should serve the same purpose now.

Reply:
removed fmtstr type.